[JSC] Heap allocation during WebAudio rendering

cdumez opened the pull request:
#### 1d11bc8cd5be50034490cc0b3e0fc569e1a7ae31
<pre>
[JSC] Heap allocation during WebAudio rendering
<a href="https://bugs.webkit.org/show_bug.cgi?id=257488">https://bugs.webkit.org/show_bug.cgi?id=257488</a>
rdar://110012510

Reviewed by Yusuke Suzuki.

Avoid a potential heap allocation under AudioWorkletGlobalScope::handlePreRenderTasks()
caused by grabbing the JSLock. Since the intention of grabbing the JSLock here is to
delay the draining of the microtasks queue until handlePostRenderTasks(), I introduce
instead a DrainMicrotaskDelayScope which delays draining of the microtasks queue without
grabbing the JSLock until the moment we want to drain the queue.

* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::drainMicrotasks):
(JSC::VM::DrainMicrotaskDelayScope::DrainMicrotaskDelayScope):
(JSC::VM::DrainMicrotaskDelayScope::~DrainMicrotaskDelayScope):
(JSC::VM::DrainMicrotaskDelayScope::operator=):
(JSC::VM::DrainMicrotaskDelayScope::increment):
(JSC::VM::DrainMicrotaskDelayScope::decrement):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::drainMicrotaskDelayScope):
* Source/WebCore/Modules/webaudio/AudioWorkletGlobalScope.cpp:
(WebCore::AudioWorkletGlobalScope::handlePreRenderTasks):
(WebCore::AudioWorkletGlobalScope::handlePostRenderTasks):
* Source/WebCore/Modules/webaudio/AudioWorkletGlobalScope.h:

Canonical link: <a href="https://commits.webkit.org/264816@main">https://commits.webkit.org/264816@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e4caa8553a79f9c6131e9d6fffbe1a1d02368740

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8395 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8688 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8905 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10062 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8457 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8404 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10676 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8789 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11319 "99 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8541 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9593 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7603 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10217 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6891 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7681 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15245 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/7213 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8012 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7824 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11180 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/8024 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8298 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6891 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/8568 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7587 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2060 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2114 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11797 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/8793 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8042 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2182 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/VM.h:
```diff
@@ -888,6 +888,24 @@ class VM : public ThreadSafeRefCounted<VM>, public DoublyLinkedListNode<VM> {
     bool enableControlFlowProfiler();
     bool disableControlFlowProfiler();
 
+    class JS_EXPORT_PRIVATE DrainMicrotaskDelayScope {
```
We would like to annotate each exposed function with JS_EXPORT_PRIVATE instead.

Constellation reviewed (approved):
r=me

cdumez committed:
[JSC] Heap allocation during WebAudio rendering
https://bugs.webkit.org/show_bug.cgi?id=257488
rdar://110012510

Reviewed by Yusuke Suzuki.

Avoid a potential heap allocation under AudioWorkletGlobalScope::handlePreRenderTasks()
caused by grabbing the JSLock. Since the intention of grabbing the JSLock here is to
delay the draining of the microtasks queue until handlePostRenderTasks(), I introduce
instead a DrainMicrotaskDelayScope which delays draining of the microtasks queue without
grabbing the JSLock until the moment we want to drain the queue.

* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::drainMicrotasks):
(JSC::VM::DrainMicrotaskDelayScope::DrainMicrotaskDelayScope):
(JSC::VM::DrainMicrotaskDelayScope::~DrainMicrotaskDelayScope):
(JSC::VM::DrainMicrotaskDelayScope::operator=):
(JSC::VM::DrainMicrotaskDelayScope::increment):
(JSC::VM::DrainMicrotaskDelayScope::decrement):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::drainMicrotaskDelayScope):
* Source/WebCore/Modules/webaudio/AudioWorkletGlobalScope.cpp:
(WebCore::AudioWorkletGlobalScope::handlePreRenderTasks):
(WebCore::AudioWorkletGlobalScope::handlePostRenderTasks):
* Source/WebCore/Modules/webaudio/AudioWorkletGlobalScope.h:

Canonical link: https://commits.webkit.org/264816@main
